samples/99_zil_interpreter/app/main.rb

```diff
@@ -7,39 +7,106 @@
 
 def tick(args)
   setup(args) if args.tick_count.zero?
-  # TODO:
-  # - Render history
-  # - Collect input
-  # - Call send_input with input when pressing Enter
+  # Render history
+  render_state(args)
+  
+  # Collect input
+  handle_input(args)
+  
+  # Call send_input when pressing Enter
+  if args.inputs.keyboard.key_down.enter && args.state.input != ''
+    send_input(args, args.state.input)
+    args.state.input = ''
+  end
+  
   $gtk.request_quit unless $interpreter.alive?
 end
 
 def setup(args)
+  args.state.text_color = [240, 255, 255]
+  args.state.bg_color = [0, 0, 0]
+  args.state.input = ''
+  args.state.text_history = []
+  args.state.current_line = 0
   context = build_zil_context(args)
   args.state.zil_context = context
   $interpreter = Fiber.new {
     context.globals[:GO].call [], context
   }
   $interpreter.resume # Initial processing until first Fiber.yield
-  process_outputs args, context.outputs # Process welcome message if existing
+  process_outputs args, ' ', context.outputs # Process welcome message if existing
+  context.outputs.clear
 
   # TODO:
   # Add other setup if necessary
 end
 
-# Call this method with the input after pressing enter
+def handle_input(args)
+  if args.inputs.keyboard.key_down.backspace
+    args.state.input.chop!
+  else
+    if args.inputs.text[0]
+      args.state.input << args.inputs.text[0]
+    end
+  end
+
+  # Scrolling
+  if args.inputs.mouse.wheel
+    args.state.current_line += args.inputs.mouse.wheel.y
+
+    if args.inputs.mouse.wheel.y > 0
+      args.state.current_line = [args.state.current_line, (args.state.text_history.length - 33)].min
+    end
+    args.state.current_line = [args.state.current_line, 0].max
+
+  # Jump to the present
+  elsif args.inputs.keyboard.key_down.escape
+    args.state.current_line = 0
+  end
+end
+
+# Called with the input after pressing enter
 def send_input(args, input)
   $interpreter.resume input
 
   context = args.state.zil_context
-  process_outputs(args, context.outputs)
+  process_outputs(args, input, context.outputs)
   context.outputs.clear
 end
 
-def process_outputs(args, outputs)
-  # TODO:
-  # Replace with actual processing, e.g. adding it to the history of output lines etc
-  outputs.each do |line|
-    puts line
+def process_outputs(args, input, outputs)
+  unless args.tick_count.zero?
+    args.state.text_history << " "
+    args.state.text_history << "> #{input}"
```

Comment:
It seems strange to me that a `process_outputs` would use inputs too ;).. maybe this can just be added at the top of send_input...

Reply:
And about what you meant in the Discord about putting `outputs.clear` into process_outputs.... maybe that's exactly what could be done..... 
we could just turn the method into 
```rb
def process_outputs(args)
  args.state.text_history += args.state.zil_context.outputs
  args.state.zil_context.outputs.clear
end
```